ODA-846 Cancer Types as json - plus a couple of tweaks

rickreesrr opened the pull request:
This Json export has some support for the upcoming story for exporting Presentation Types as Json. At the moment we don't have enough information to confirm the format.
There is a small tweak to stop the GelJsonExporter from recursing too far to speed it up.
There's a change in the controller to get the latest version for RareDiseases doc export. I think this will now change to use the metadata solution though.

musketyr commented on ModelCatalogueCorePlugin/src/groovy/org/modelcatalogue/core/util/Metadata.groovy:
```diff
@@ -14,5 +14,6 @@ class Metadata {
      * The HPO and clinical tests report exporters only work for and should only be available to the relevant data class(es), discerned by this metadata field
      */
     public static final def HPO_REPORT_AVAILABLE = "rare-disease-hpo-clinical-tests-report"
+    public static final def CANCER_TYPES_AVAILABLE = "cancer-types-export"
```
this applies to the HPO_REPORT_AVAILABLE, it's better to create the metadata keys as URLs where can be some explanation put in future

rickreesrr replied:
Don't understand how URLs help, could you explain further. Does this need to change from what I've submitted?

musketyr replied:
from my point of view
1. URL gives the user better understanding that this metadata value is something special and it is going to be processed by the catalogue
2. minimize the name clash and applying this flag by accident
3. if the URL points to the webpage the webpage can contain more information how does the flag work
4. this is a way how metadata with special meaning are declared across the catalogue (see e.g. CRF exporter or new metadata on data model)


I can live with what you've submitted but to be coherent across the catalogue I would prefer to change the key to some url like `http://www.genomicsengland.co.uk/metadata/cancer-types-export` or similar.

musketyr commented on ModelCatalogueGenomicsPlugin/test/integration/org/modelcatalogue/gel/export/CancerTypesExporterSpec.groovy:
```diff
@@ -0,0 +1,186 @@
+ * Created by rickrees on 31/03/2016.
+ */
+class CancerTypesExporterSpec extends IntegrationSpec {
+    public static final String DATA_MODEL_NAME = 'testDataModel1'
```
very elegant test, I like it :+1:

musketyr commented on ModelCatalogueGenomicsPlugin/grails-app/controllers/org/modelcatalogue/core/genomics/GenomicsController.groovy:
```diff
@@ -146,23 +145,24 @@ class GenomicsController {
-        DataClass model = DataClass.get(params.id)
+        DataClass dClass = DataClass.get(params.id)
+
+        DataClass latestVersion = (DataClass) elementService.findByModelCatalogueId(DataClass, dClass.getDefaultModelCatalogueId(true))
```
I can see where you're heading but this may confuse users a lot. what if you want to export older version and you will always get the latest?

musketyr commented:
apart from the latest version hacks looks very good. I'll wait with the merge until the latest version hack is cleared.

rickreesrr commented:
Fixed the issues you flagged up but complicated your merge by adding the CSV export functionality into this branch (sorry!). I refactored the test pulling out an AbstractCancerTypesExporterSpec as I wanted to re-use the complex model for both CSV & JSON exporters. And it will hopefully be used by the Presentation type stories coming up.

musketyr commented:
looks good to me. thanks for the changes :+1:
